Gamification

AlfioEmanueleFresta opened the pull request:
Adds gamification functionalities.

AlfioEmanueleFresta committed:
Created QuizResult model class to store student results for quizzes

AlfioEmanueleFresta committed:
Merge branch 'master' of github.com:DAMS-Bale/flippd into Gamification

JGinn committed:
Store The Quiz Results After Marking

Modify the route such that it stores the results for a quiz after
it has been marked. Create a test to verify that this occurs.

AlfioEmanueleFresta committed:
Ref #9, created user profile page

- Quiz#number_of_correct_answer to get the total no of correct answers
- Created the user page view, which includes
  - Latest results bar chart using google charts
  - Best results table

AlfioEmanueleFresta committed:
Ref #9, implemented leaderboard

- Leaderboard showing best performing users
- Fixed the best results query for the users
- 'No results' message for user pages with no results yet
- Hidden the results chart for users with no results yet

AlfioEmanueleFresta committed:
Ref #9, button to toggle sharing results

In the leaderboard, users can now switch results sharing on and off using
 a toggle switch.

AlfioEmanueleFresta committed:
Whitespaces

AlfioEmanueleFresta committed:
Ref #9 added tests for the leaderboard

- The helper method sign_in now accepts optional `name` and `email` arguments
- Created tests for the opt-in privacy switch
- Created tests for the correct score being displayed in the leaderboard

JGinn committed:
Add Page View Tracking

* Add tracking of page views for logged in users for video pages.
* Add the viewed videos to the dashboard.
* Add tests of this functionality.
* Seperate video route into a new file.

JGinn committed:
Merge branch 'Gamification' of https://github.com/DAMS-Bale/flippd into Gamification

JGinn committed:
Restore Empty Results Messages

Restore the empty results messages for #7 on a user's dashboard page.

JGinn committed:
Change Page Views to Video Views

Change page views to video views, for increased clarity, as generic
page view functionality is not being used, so is not required.

AlfioEmanueleFresta committed:
Leaderboard moved to Leaderboard model

AlfioEmanueleFresta committed:
Ref #9, Leaderboard refactoring
- Moved leaderboard to Leaderboard model
- Results are inserted from User#add_result, which checks whether the result
  is the new best for the quiz or not, and stores in the QuizResult#best property.
- User#best_results now returns QuizResult objects and does not use SQL
- Leaderboard#results now returns a sorted list of hashes ({:user, :score})
- The application is now SQL-free!

AlfioEmanueleFresta committed:
Ref #9, fixed best-score logic

JGinn committed:
Add Video View Points

Add points for viewing videos to the total score for a user, used
for calculating the leaderboard values.

JGinn committed:
Add Points to Dashboard, Test Points from Videos

For ref #9.

JGinn committed:
Add Tests of Quiz Results on Dashboard

For ref #9, add testing that the best quiz result is shown on the
dashboard.

JGinn committed:
Test Points Awarded For Quizzes

Add tests that verify that points are awarded for completing
quizzes, and stored on the dashboard. For ref #9.

AlfioEmanueleFresta commented on app/models/user.rb:
```diff
@@ -21,4 +25,50 @@ def can_edit_comment comment
+      total += result.score
+    end
+    # Add the score for watching videos.
+    total += self.video_views.size * 5
```
@JGinn thanks! I completely forgot about that.

JGinn replied:
Removed the extra line.

AlfioEmanueleFresta commented:
:+1: Looks good to me!

JGinn commented on app/routes/quiz.rb:
```diff
@@ -8,7 +8,18 @@ class Flippd < Sinatra::Application
+    # Only save if logged in.
+    if @user then
+
+      quiz = Quiz.get(params[:id])
```
Does this line do nothing, since line 11 is always called before?

AlfioEmanueleFresta replied:
It appears so, my mistake.

JGinn replied:
No worries. I'll remove this line and simplify the loop in routes/video.rb then update the PR.

JGinn commented on app/routes/video.rb:
```diff
@@ -0,0 +1,43 @@
+    @phases.each do |phase|
+      phase['topics'].each do |topic|
+        topic['videos'].each do |video|
+          if video["id"] == params['id'].to_i
```
We could combine the contents of the loops so that there is only one loop, then an if id... else if id + 1... else if id -1...
EDIT: Done

JGinn committed:
Clean Up

For ref #9, clean up extra duplicated line from quiz route, and
simplify the video route.

JGinn commented:
All seems good to me now, merging.
